src/gauging.jl

```diff
@@ -0,0 +1,58 @@
+"""Modify the diagonal elements of an ITensor with a specified function"""
+function modify_diagonal_els_itensor(A::ITensor, f::Function)
+  out = copy(A)
+  is = [[j for i in 1:length(inds(A))] for j in 1:mindim(out)]
+  for i in is
+    out[i...] = f(out[i...])
+  end
+
+  return out
+end
+
+"""Put an ITensorNetwork into the symmetric gauge and also return the message tensors (which are the diagonal bond matrices)"""
+function symmetrise_itensornetwork(ψ::ITensorNetwork)
+  ψψ = ψ ⊗ prime(dag(ψ); sites=[])
+  vertex_groups = nested_graph_leaf_vertices(partition(ψψ, group(v -> v[1], vertices(ψψ))))
+  mts = compute_message_tensors(ψψ; vertex_groups=vertex_groups)
+
+  ψsymm = copy(ψ)
+  symm_mts = copy(mts)
+
+  for e in edges(ψ)
+    vsrc, vdst = src(e), dst(e)
+
+    s1, s2 = find_subgraph((vsrc, 1), mts), find_subgraph((vdst, 1), mts)
+    edge_ind = commoninds(mts[s1 => s2], ψsymm[vsrc])
+    edge_ind_sim = sim(edge_ind)
+
+    X_D, X_U = eigen(mts[s1 => s2]; ishermitian=true)
+    Y_D, Y_U = eigen(mts[s2 => s1]; ishermitian=true)
+
+    rootX_D, rootY_D = modify_diagonal_els_itensor(X_D, x -> x^0.5),
+    modify_diagonal_els_itensor(Y_D, x -> x^0.5)
+    inv_rootX_D, inv_rootY_D = modify_diagonal_els_itensor(X_D, x -> x^-0.5),
+    modify_diagonal_els_itensor(Y_D, x -> x^-0.5)
+    rootX = X_U * rootX_D * prime(dag(X_U))
+    rootY = Y_U * rootY_D * prime(dag(Y_U))
+    inv_rootX = X_U * inv_rootX_D * prime(dag(X_U))
+    inv_rootY = Y_U * inv_rootY_D * prime(dag(Y_U))
+
+    ψsymm[vsrc] = noprime(ψsymm[vsrc] * inv_rootX)
+    ψsymm[vdst] = noprime(ψsymm[vdst] * inv_rootY)
+
+    Ce = rootX * replaceinds(rootY, edge_ind, edge_ind_sim)
+
+    U, S, V = svd(Ce, edge_ind)
+    rootS = modify_diagonal_els_itensor(S, x -> x^0.5)
+
+    ψsymm[vsrc] = replaceinds(ψsymm[vsrc] * U * rootS, commoninds(S, V), edge_ind)
+    ψsymm[vdst] = replaceinds(ψsymm[vdst], edge_ind, edge_ind_sim)
+    ψsymm[vdst] = replaceinds(ψsymm[vdst] * dag(rootS * V), commoninds(U, S), edge_ind)
+
+    replaceinds!(S, inds(S), inds(mts[s1 => s2]))
```

Comment:
Let's use:
```julia
S = replaceinds(S, inds(S) => inds(mts[s1 => s2]))
```
(I wish we didn't have the in-place version, there is essentially no performance advantage and it can lead to confusing bugs if you accidentally modify the indices of a reference to the same ITensor).

Though this kind of code is a bit suspicious to me, I we certain we always replace the indices in the correct way (i.e. with the correct ordering?). I guess because `S` is diagonal it is relatively safe but the code above using `replaceinds` definitely raises alarm bells for me.

Reply:
Yes because `S` is diagonal I didn't think much about the ordering of the indices but maybe that is a bit sloppy. It is not clear to me which index of `S` should be the `edge_ind` and which should be `prime(edge_ind)` though? Perhaps we could just be a bit more precise and use `replaceinds(S, [commonind(U, S), commonind(V,S)] => [edge_ind, prime(edge_ind)]`?

Reply:
Yes, that would be a good solution. At least then it is explicit when someone reads the code, even if there are no practical issues because it is diagonal.